Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/exception/Backtrace.java

```diff
@@ -0,0 +1,225 @@
+         *         function call arguments.
+         */
+        public List<Variable> getArguments() {
+            return this.arguments;
+        }
+
+        /**
+         * Returns the local variable bindings in the scope of
+         * expression evaluation of the frame.
+         *
+         * @return A list of the {@link Variable}s that correspond to
+         *         local variable bindings.
+         */
+        public List<Variable> getLocalVariables() {
+            return this.localVariables;
+        }
+
+        /**
+         * Returns the name of the CQL context that was current during
+         * the evaluation represented by the frame.
+         *
+         * @return The name of the CQL context.
+         */
+        public String getContextName() {
+            return this.contextName;
+        }
+
+        /**
+         * Returns the value of the CQL context that was current
+         * during the evaluation represented by the frame.
+         *
+         * @return The value of the CQL context.
+         */
+        public Object getContextValue() {
+            return this.contextValue;
+        }
+    }
+
+    private final List<Frame> frames = new LinkedList<>();
+
+    public List<Frame> getFrames() {
+        return this.frames;
+    }
+
+    public void addFrame(final Frame frame) {
+        this.frames.add(frame);
+    }
+
+    public void maybeAddFrame(
+            final ExpressionDef containingDefinition,
+            final State.ActivationFrame definitionFrame,
+            final Deque<State.ActivationFrame> stack,
+            final String contextName,
+            final Object contextValue,
+            final Expression expression) {
+        // When the EvaluationVisitor unwinds through
+        // EvaluationVisitor.visitExpression calls, every call has a
+        // unique expression that is being evaluated but not every
+        // call has a unique ActivationFrame since each
+        // ActivationFrame is associated with the respective innermost
+        // expression definition or function definition enclosing the
+        // expression which is being evaluated in the call. For that
+        // reason, this method is called for multiple expressions with
+        // the same surrounding ActivationFrame. If the frame at the
+        // end of our frame list already describes the definition or
+        // function that surrounds the expression for which the
+        // visitExpression is being unwound, everything has already
+        // been recorded: the innermost expression, the containing
+        // definition or function and all local variables and
+        // arguments.
+        if (!frames.isEmpty()) {
+            final var currentFrame = this.frames.get(frames.size() - 1);
+            if (currentFrame instanceof FunctionoidFrame) {
+                if (((FunctionoidFrame) currentFrame).getDefinition() == containingDefinition) {
+                    return;
+                }
+            }
```

Comment:
SonarQube flags this as a maintainability issue, so:
```suggestion
            if (currentFrame instanceof FunctionoidFrame functionoidFrame) {
                if (functionoidFrame.getDefinition() == containingDefinition) {
                    return;
                }
            }
```